src/Workspaces/Core/Portable/FlowAnalysis/SymbolUsageAnalysis/SymbolUsageAnalysis.BasicBlockAnalysisData.cs

```diff
@@ -0,0 +1,178 @@
+            }
+
+            /// <summary>
+            /// Gets the currently reachable writes for the given symbol.
+            /// </summary>
+            public IEnumerable<IOperation> GetCurrentWrites(ISymbol symbol)
+            {
+                if (_reachingWrites.TryGetValue(symbol, out var values))
+                {
+                    foreach (var value in values)
+                    {
+                        yield return value;
+                    }
+                }
+            }
+
+            /// <summary>
+            /// Marks the given symbol write as a new unread write operation,
+            /// potentially clearing out the prior write operations if <paramref name="maybeWritten"/> is <code>false</code>.
+            /// </summary>
+            public void OnWriteReferenceFound(ISymbol symbol, IOperation operation, bool maybeWritten)
+            {
+                if (!_reachingWrites.TryGetValue(symbol, out var values))
+                {
+                    values = PooledHashSet<IOperation>.GetInstance();
+                    _reachingWrites.Add(symbol, values);
+                }
+                else if (!maybeWritten)
+                {
+                    values.Clear();
+                }
+
+                values.Add(operation);
+            }
+
+            public bool Equals(BasicBlockAnalysisData other)
+                => this.GetHashCode() == (other?.GetHashCode() ?? 0);
+
+            public override bool Equals(object obj)
+            {
+                return Equals(obj as BasicBlockAnalysisData);
+            }
+
+            public override int GetHashCode()
+            {
+                var hashCode = _reachingWrites.Count;
+                foreach ((int keyHash, int valueHash) in _reachingWrites.Select(GetKeyValueHashCodeTuple).OrderBy(hashTuple => hashTuple.keyHash))
+                {
+                    hashCode = Hash.Combine(Hash.Combine(keyHash, valueHash), hashCode);
+                }
+
+                return hashCode;
+
+                // Local functions.
+                (int keyHash, int valueHash) GetKeyValueHashCodeTuple(KeyValuePair<ISymbol, PooledHashSet<IOperation>> keyValuePair)
+                    => (keyValuePair.Key.GetHashCode(), Hash.Combine(keyValuePair.Value.Count, Hash.CombineValues(keyValuePair.Value)));
+            }
+
+            private bool IsEmpty => _reachingWrites.Count == 0;
+
+            public static BasicBlockAnalysisData Merge(
+                BasicBlockAnalysisData data1,
+                BasicBlockAnalysisData data2,
+                Func<BasicBlockAnalysisData> createBasicBlockAnalysisData)
+            {
+                if (data1 == null)
+                {
+                    return data2;
+                }
+                else if (data2 == null)
+                {
+                    return data1;
+                }
+                else if (data1.IsEmpty)
+                {
+                    return data2;
```

Comment:
nit, consider: `if (data == null || data.IsEmpty) ...`

Reply:
If anyone of them is null, we want to return the non-null instance.

---
In reply to: [231670257](https://github.com/dotnet/roslyn/pull/31025#discussion_r231670257) [](ancestors = 231670257)

Reply:
ah, interesting.  thanks!  possibly worth documenting that that's important to perserve.